main.go

```diff
@@ -0,0 +1,543 @@
+// task. Returns true if it successfully claimed a task (regardless of whether
+// the task ran successfully) otherwise false.
+func FindAndRunTask() bool {
+	// Write to the signed urls channel, to request signed urls back on
+	// channel c.
+	signedURLsRequestChan <- signedURLsResponseChan
+	// Read the result.
+	signedURLs := <-signedURLsResponseChan
+	taskFound := false
+	// Each of these signedURLs represent an underlying Azure queue, there
+	// are multiple of these so that we can support priority. For this
+	// reason the worker must poll the Azure queues in order they are
+	// given.
+	for _, urlPair := range signedURLs.Queues {
+		// try to grab a task using the url pair (url pair = poll url + delete
+		// url)
+		task, err := SignedURLPair(urlPair).Poll()
+		if err != nil {
+			// This can be any error at all occurs in queryAzureQueue that
+			// prevents us from claiming this task.  Log, and continue.
+			log.Println(err)
+			continue
+		}
+		// Now we found a task, run it, and then exit the loop. This is because
+		// the loop is in order of priority, most important first, so we will
+		// run the most important task we find, and then return, ignorning
+		// remaining urls for lower priority tasks that might still be left to
+		// loop through, since by the time we complete the first task, maybe
+		// higher priority jobs are waiting, so we need to poll afresh.
+		log.Println("Task found")
+		taskFound = true
+		// If there is one or more messages the worker must claim the tasks
+		// referenced in the messages, and delete the messages.
+		err = task.claim()
+		if err != nil {
+			log.Printf("WARN: Not able to claim task %v\n", task.TaskId)
+			log.Println(err)
+			continue
+		}
+		task.setReclaimTimer()
+		err = task.fetchTaskDefinition()
+		if err != nil {
+			log.Printf("WARN: Not able to fetch task definition for task %v\n", task.TaskId)
+			log.Println(err)
+			continue
+		}
+		err = task.validatePayload()
+		if err != nil {
+			log.Printf("WARN: Not able to validate task payload for task %v\n", task.TaskId)
+			log.Println(err)
+			continue
+		}
+		err = task.run()
+		if err != nil {
+			log.Printf("WARN: Not able to run task %v\n", task.TaskId)
+			log.Println(err)
+			continue
+		}
+		break
+	}
+	return taskFound
+}
+
+// Queries the given Azure Queue signed url pair (poll url/delete url) and
+// translates the Azure response into a Task object
+func (urlPair SignedURLPair) Poll() (*TaskRun, error) {
+	queueMessagesList := new(QueueMessagesList)
+	// To poll an Azure Queue the worker must do a `GET` request to the
+	// `signedPollUrl` from the object, representing the Azure queue. To
+	// receive multiple messages at once the parameter `&numofmessages=N`
+	// may be appended to `signedPollUrl`. The parameter `N` is the
+	// maximum number of messages desired, `N` can be up to 32.
+	// Since we can only process one task at a time, grab only one.
+	resp, err := http.Get(urlPair.SignedPollUrl + "&numofmessages=1")
```

Comment:
> Do you agree with this choice? I could be oversimplifying, let me know what you think.
I think the simplification will work...

Personally I would do retries with exponential backoff even if it slows down, because can very well be desired result...
Also maybe polling every 5-10s is better... Or just make it configurable.

Just make sure all other API calls are retried... Hence, you might as well implement a generic method for retrying stuff. You can probably do a retryFunction(callback, TRANSIENT_ERRORS) function where callback is a function that does the request and returns any error... and TRANSIENT_ERRORS is a list of transient error codes...
Mostly because you'll critically need it for the delete message from azure call..

Note, the go taskcluster client library should do the retries already (right), if not we should definitely fix that first.